lib/translator/inquiry.cc

```diff
@@ -0,0 +1,247 @@
+    absl::Span<uint8_t> buffer) {
+  scsi_defs::InquiryData result = {
+      .version = scsi_defs::Version::kSpc4,
+      .response_data_format = scsi_defs::ResponseDataFormat::kCompliant,
+      .additional_length = 0x1f,
+      .tpgs = scsi_defs::TPGS::kNotSupported,
+      .protect = (identify_namespace_data.dps.pit == 0 &&
+                  identify_namespace_data.dps.md_start == 0)
+                     ? 0
+                     : 1,
+      .cmdque = 1};
+
+  // Shall be set to “NVMe" followed by 4 spaces: “NVMe    “
+  // Vendor Identification is not null terminated.
+  static_assert(sizeof(result.vendor_identification) ==
+                kNvmeVendorIdentification.size());
+
+  memcpy(result.vendor_identification, kNvmeVendorIdentification.data(),
+         kNvmeVendorIdentification.size());
+
+  // Shall be set to the first 16 bytes of the Model Number (MN) field within
+  // the Identify Controller Data Structure
+  memcpy(result.product_identification, identify_controller_data.mn,
+         sizeof(result.product_identification));
+
+  /*
+  Shall be set to the last 4 ASCII graphic characters in the range of 21h-7Eh
+  (i.e. last 4 non-space characters) of the Firmware Revision (FR) field within
+  the Identify Controller Data Structure.
+  */
+  int idx = 3;
+  for (int i = 7; i >= 0; i--) {
+    if (identify_controller_data.fr[i] != ' ' &&
+        identify_controller_data.fr[i] >= 0x21 &&
+        identify_controller_data.fr[i] <= 0x7e) {
+      result.product_revision_level[idx] = identify_controller_data.fr[i];
+      if (idx-- == 0) break;
+    }
+  }
+
+  memcpy(buffer.data(), &result, sizeof(result));
+}
+
+void TranslateSupportedVpdPages(absl::Span<uint8_t> buf) {
+  scsi_defs::PageCode supported_page_list[7] = {
+      scsi_defs::PageCode::kSupportedVpd,
+      scsi_defs::PageCode::kUnitSerialNumber,
+      scsi_defs::PageCode::kDeviceIdentification,
+      scsi_defs::PageCode::kExtended,
+      scsi_defs::PageCode::kBlockLimitsVpd,
+      scsi_defs::PageCode::kBlockDeviceCharacteristicsVpd,
+      scsi_defs::PageCode::kLogicalBlockProvisioningVpd};
+
+  scsi_defs::SupportedVitalProductData result = {
+      .page_length = sizeof(supported_page_list),
+  };
+
+  memcpy(buf.data(), &result, sizeof(result));
+  memcpy(&buf[sizeof(result)], &supported_page_list,
+         sizeof(supported_page_list));
+}
+
+void TranslateUnitSerialNumberVpd(
+    const nvme_defs::IdentifyControllerData &identify_controller_data,
+    const nvme_defs::IdentifyNamespace &identify_namespace_data, uint32_t nsid,
+    absl::Span<uint8_t> buffer) {
+  scsi_defs::UnitSerialNumber result = {
+      .page_code = scsi_defs::PageCode::kUnitSerialNumber};
+
+  // converts the nguid or eui64 into a formatter hex string
+  // 0x0123456789ABCDEF would be converted to “0123_4567_89AB_CDEF."
+  char product_serial_number[100] = {0}, hex_string[33];
+  const size_t kNguidLen = 40, kEui64Len = 20, kV1SerialLen = 30;
```

Comment:
Something worth considering: we could define these as constants in the header file, so we could reuse the constant in your test code.
Instead of testing `ASSERT_EQ(result.page_length, 40);` we could have `ASSERT_EQ(result.page_length, kNguidLen);`, and so forth.
This makes clear the relationship between the line of code that sets the value, and the line of code that tests it.

Reply:
will change this when inquiry branches are merged together -- think the code for this is already there in a diff branch

Reply:
added a TODO